Allows usage of public ip, private ip or hostname in Linode inventory

mcg opened the pull request:
The older linode inventory plugin supported using the public IP as well as the
instances label. This change enables using public IP, private IP or hostname. It also injects public_ip and private_ip into `hostvars` for easy access(also as the old plugin).

##### SUMMARY
Fixes #51195 

##### ISSUE TYPE
- Feature Pull Request


##### COMPONENT NAME
`lib/ansible/plugins/inventory/linode.py`

mcg committed:
Brings back using public ip instead of label in Linode inventory

The older linode inventory plugin supported using the public_ip as well as the
instances label. This re-enabled this functionality along with allowing
private_ip. It also injects both public_ip and private_ip into `hostvars`. As
the older inventory plugin did.

Acked-by: Matthew Gregg <matthew.gregg@modernmsg.com>

Akasurde commented:
CI failure -  https://app.shippable.com/github/ansible/ansible/runs/103323/6/console
```
00:18 =================================== FAILURES ===================================
00:18 _______________________ test_empty_config_query_options ________________________
00:18 [gw1] linux2 -- Python 2.7.12 /usr/bin/python2.7
00:18 inventory = <ansible.plugins.inventory.linode.InventoryModule object at 0x7f45f85cb950>
00:18 
00:18     def test_empty_config_query_options(inventory):
00:18 >       regions, types = inventory._get_query_options({})
00:18 E       AttributeError: 'InventoryModule' object has no attribute '_get_query_options'
00:18 
00:18 test/units/plugins/inventory/test_linode.py:61: AttributeError
00:18 ___________________________ test_conig_query_options ___________________________
00:18 [gw1] linux2 -- Python 2.7.12 /usr/bin/python2.7
00:18 inventory = <ansible.plugins.inventory.linode.InventoryModule object at 0x7f45f85cb950>
00:18 
00:18     def test_conig_query_options(inventory):
00:18 >       regions, types = inventory._get_query_options({
00:18             'regions': ['eu-west', 'us-east'],
00:18             'types': ['g5-standard-2', 'g6-standard-2'],
00:18         })
00:18 E       AttributeError: 'InventoryModule' object has no attribute '_get_query_options'
00:18 
00:18 test/units/plugins/inventory/test_linode.py:66: AttributeError
00:18 --- generated xml file: /root/ansible/test/results/junit/python2.7-units.xml ---
00:18 =========================== short test summary info ============================
00:18 FAIL test/units/plugins/inventory/test_linode.py::test_empty_config_query_options
00:18 FAIL test/units/plugins/inventory/test_linode.py::test_conig_query_options
```

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,31 @@ def _filter_by_config(self, regions, types):
         for instance in self.instances:
-            self.inventory.add_host(instance.label, group=instance.group)
+            if use_public_ip:
+                dest = self._get_instance_public_ip(instance)
```
Can we name `dest` better? I feel like `host_identifier` (or so) might be more readable.

mcg replied:
Will change

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,31 @@ def _filter_by_config(self, regions, types):
+            self.inventory.add_host(dest, group=instance.group)
+
+    def _get_instance_public_ip(self, instance):
+        """Returns a the public IP address of the instance"""
```
`%s/a the/the`

mcg replied:
Oops. Will fix.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,31 @@ def _filter_by_config(self, regions, types):
+
+    def _get_instance_public_ip(self, instance):
+        """Returns a the public IP address of the instance"""
+        return [ip for ip in instance.ipv4 if not ipaddress.ip_address(ip).is_private][0]
```
I think we should guard against an `IndexError`. Even if Linode guarantee us some value, given programming error and such, I'd like to guarantee users of this module error messages that lead them in the right direction.

mcg replied:
Added some guarding here. Hopefully what you're going for.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,31 @@ def _filter_by_config(self, regions, types):
         """Add hostvars for instances in the dynamic inventory."""
         for instance in self.instances:
+            if use_public_ip:
+                dest = self._get_instance_public_ip(instance)
```
Again, can we name this `dest` better? Will help readability as the module grows.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -180,13 +203,14 @@ def _get_query_options(self, config_data):
         """Verify the Linode configuration file."""
         if super(InventoryModule, self).verify_file(path):
-            endings = ('linode.yaml', 'linode.yml')
+            endings = ('.linode.yaml', '.linode.yml')
```
Did you experience some bug with the `endings` detection or?

mcg replied:
No, just my local env out of sync. Fixed.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -158,8 +177,8 @@ def _validate_option(self, name, desired_type, option_value):
 
-    def _get_query_options(self, config_data):
-        """Get user specified query options from the configuration."""
+    def _get_user_options(self, config_data):
```
I'm not sure why you changed this? It could be described both ways, I suppose. 

Hence, this feels like it could be bikeshed :sweat_smile:.

mcg replied:
"query" felt like this was only related to configuring the API query, which it was with region/type. Adding `use_public_ip` isn't related to query, just config.

decentral1se replied:
Got it, agreed!

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -200,9 +224,9 @@ def parse(self, inventory, loader, path, cache=True):
         self._add_groups()
-        self._add_instances_to_groups()
-        self._add_hostvars_for_instances()
+        self._add_instances_to_groups(use_public_ip)
```
Thinking ahead, in case there are other ways to specify the identifier for an instance (lable, ip, etc.), we might implement here a `self._determine_host_identifier` or so, which negotiates the options and decides if we're using the `public_ip` or the `label` and then passes that to the `_add_instances_to_groups` and `_add_hostvars_for_instances`? We could then eliminate the `if/else` clause in those function sand keep the control flow more simple. WDYT?

decentral1se reviewed (changes requested):
Thanks for this @mcg! I've asked other Linode working group members to review and keep an eye on this, so we should be getting it in promptly. We're lurking on freenode/#ansible-linode if you have any questions. Otherwise, just ping here.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,39 @@ def _filter_by_config(self, regions, types):
+                    return ip
+            return False
+        else:
+            raise IndexError(
```
This should use `AnsibleError` like the other errors to pass back a nicely formatted message to the user.

mcg replied:
Wasn't sure about using that here, will change.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,39 @@ def _filter_by_config(self, regions, types):
+            for ip in instance.ipv4:
+                if not ipaddress.ip_address(ip).is_private:
+                    return ip
+            return False
```
I don't understand why we would `return False` here. This function is about giving back an instance based on the public IP address or failing gracefully with an error message for the user. If this `return False` is happening, we need to write code to handle that in all places it is called within the module. This can lead to extra `if/else` which we should avoid for maintainability purposes.

mcg replied:
Hmm ok. Will rework this.

decentral1se commented on test/units/plugins/inventory/test_linode.py:
```diff
@@ -58,18 +58,21 @@ def test_validation_option_bad_option(inventory):
 
-def test_conig_query_options(inventory):
-    regions, types = inventory._get_query_options({
+def test_conig_user_options(inventory):
```
Oh, we should `%s/conig/config/`

mcg replied:
sigh.. will fix

decentral1se reviewed (changes requested):
Thanks for pushing on with this @mcg. I've given it another pass.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +131,45 @@ def _filter_by_config(self, regions, types):
+                host_identifier = instance.label
+            self.inventory.add_host(host_identifier, group=instance.group)
+
+    def _get_instance_public_ip(self, instance):
```
Could we get a few tests for the different flows in this function?

decentral1se reviewed (changes requested):
Final one! :fire: :fire: :fire:

decentral1se commented:
Bump! Any news on this @mcg? Thanks for the efforts :100:

mcg commented:
@decentral1se Trying to get back to this, what we're using internally has diverged a bit from this work. Basically, instead of `use_public_ip: true/false`, we're using `instance_access: public_ip/private_ip/hostname` along with injecting `public_ip` and `private_ip` into the inventory as the old plugin also did. Should I push this work to this PR?

decentral1se commented:
Hey @mcg, I'm not sure I follow, does that map what the Linode API tooling exposes with v4? If so, then sure, I'd be happy to review and push to get merged. Not sure if we want adhoc keys otherwise - but please tell if I have this wrong.

mcg commented:
@decentral1se the old plugin added keys for public_ip and private_ip(and a few other values), so that you can access and use these in hostvars directly. They're not adhoc so much as just not directly accessible in the API. This is something we use heavily for building firewall rules, loadbalancer backends etc...

Here is what we're doing:
```
    def _add_hostvars_for_instances(self, instance_access):
        """Add hostvars for instances in the dynamic inventory."""
        for instance in self.instances:
            if instance_access == 'public_ip':
                host_identifier = self._get_instance_public_ip(instance)
            elif instance_access == 'private_ip':
                host_identifier = self._get_instance_private_ip(instance)
            elif instance_access == 'hostname':
                host_identifier = instance.label
            else:
                raise AnsibleError(
                    'Instance Access setting is not valid'
                )

            self.inventory.set_variable(
                host_identifier,
                "public_ip",
                self._get_instance_public_ip(instance)
            )
            self.inventory.set_variable(
                host_identifier,
                "private_ip",
                self._get_instance_private_ip(instance)
            )
            hostvars = instance._raw_json
            for hostvar_key in hostvars:
                self.inventory.set_variable(
                    host_identifier,
                    hostvar_key,
                    hostvars[hostvar_key]
                )
```

decentral1se commented:
Oh yes, that looks powerful and useful, let's have it! 

You can rename this PR + wipe the changes or open another, as you want. 

Just ping and I'll happily review. :rocket:

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,85 @@ def _filter_by_config(self, regions, types):
+
+            self.inventory.set_variable(
+                host_identifier,
+                "public_ip",
```
```suggestion
                'public_ip',
```

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,85 @@ def _filter_by_config(self, regions, types):
+            )
+            self.inventory.set_variable(
+                host_identifier,
+                "private_ip",
```
```suggestion
                'private_ip',
```

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,85 @@ def _filter_by_config(self, regions, types):
             ]
 
-    def _add_instances_to_groups(self):
+    def _get_instance_public_ip(self, instance):
```
We could merge the two `_get_instance*_ip` functions with something like?

```
def _get_instance_by_ip(self, instance, is_private=False):
    """Returns the IPv4 address of the instance."""
    if instance.ipv4 and len(instance.ipv4) > 0:
        for ip in instance.ipv4:
            parsed_ip = ipaddress.ip_address(ip)

            if parsed_ip.is_private and is_private:
                return ip

            if parsed_ip.is_private and not is_private
                raise AnsibleError('Expected public IP not found')

            return parsed_ip

    raise AnsibleError('Instance IPv4 address is missing or empty')
```

Less code :rocket:

mcg replied:
How about....
```
  def _get_instance_ip(self, instance, private=False):
        """Returns the public or private IP address of the instance"""
        if instance.ipv4 and len(instance.ipv4) > 0:
            result_ip = None
            for ip in instance.ipv4:
                if not ipaddress.ip_address(ip).is_private and not private:
                    result_ip = ip
                elif ipaddress.ip_address(ip).is_private and private:
                    result_ip = ip
            if result_ip:
                return result_ip
        else:
            raise AnsibleError(
                'Instance IPv4 address is missing or empty'
            )
```

decentral1se replied:
Yes, great!

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -29,6 +29,11 @@
             env:
                 - name: LINODE_ACCESS_TOKEN
+        instance_access:
+          description: How to access the instance(hostname, public_ip, private_ip).
```
You can use the following docs key to specify what choices are valid:

```
choices:
- hostname
- public_ip
- private_ip
```

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,85 @@ def _filter_by_config(self, regions, types):
+            elif instance_access == 'hostname':
+                host_identifier = instance.label
+            else:
+                raise AnsibleError(
```
If we use `choices` in the documentation fragment, we can avoid this (invalid values will never be allowed here).

decentral1se reviewed (changes requested):
Thanks for picking this up again :+1:

mcg committed:
Adds tests for get public and private ip's

mcg committed:
Combines get public/private into a single function

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,61 @@ def _filter_by_config(self, regions, types):
+                    result_ip = ip
+                elif ipaddress.ip_address(ip).is_private and private:
+                    result_ip = ip
+            if result_ip:
```
Oh woops, are we missing that `raise AnsibleError...` when the `result_ip` is `None`?

mcg replied:
I keep forgetting private ip is optional. Added that back and a test.

decentral1se commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -125,18 +132,61 @@ def _filter_by_config(self, regions, types):
+                host_identifier = self._get_instance_ip(instance, private=True)
+            elif instance_access == 'hostname':
+                host_identifier = instance.label
+            else:
```
Same as https://github.com/ansible/ansible/pull/51196#discussion_r266537178, we can drop it?

mcg replied:
Drop what? Even using "choice" we have to set the `host_identifier` based on the choice?

decentral1se replied:
Ah, right, got it :+1:

decentral1se commented:
Final comments, promise :) Looks good!

mcg committed:
Handle missing private ip and associated test

decentral1se reviewed (approved):
:rocket: 

shipit

mcg commented:
@decentral1se Thanks for all the help!

decentral1se commented:
@InTheCloudDan @displague @rmcintosh, can we get some shipit gold on this one?

InTheCloudDan commented:
shipit

decentral1se commented:
Looks like `core_review` shouldn't be marked for this issue (see https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#core). I've gone and nagged on `#ansible-devel` to get it pushed through.

gundalow commented:
bot_status
Wondering where `support:core` is being set

rmcintosh commented:
shipit

bcoca commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -29,6 +29,11 @@
             required: true
             env:
                 - name: LINODE_ACCESS_TOKEN
+        instance_access:
```
Instead of hardcoding this choice, why not use constructed/compose functionality to allow user to set `ansible_host` and similar variables to anything they need.

mcg replied:
Can you explain a bit more how this would work? I've dug around a bit and am not sure.

bcoca replied:
basically #53869 does same for vultr plugin

mcg replied:
Does make sense to use `ansible_host` rather than change the host identifier. I'll make that change.

bcoca replied:
yes, since that is what you specifically are doing with the new option 'setting ansible_host' while the compose system gives you a generic way of setting any/all kinds of variables, you really only need this type of option for 'inventory_hostname'

mcg replied:
Thanks for following up, this does use `ansible_host` now.

bcoca replied:
maybe I am not being clear, my point is you do NOT need THIS specific option, using compose any user can decide exactly what `ansible_host` value is set to, as well as any other connection or non connection specific variable.

decentral1se commented:
FYI @mcg, you've got until tomorrow to get this merged or it will be frozen out due to https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst. No pressure :)

mcg committed:
Injects ansible_host, rather than change the instance id.

 - Uses `label` rather than `hostname` to match linode naming

mcg commented:
When is the next window to get this merged, after 2019-05-16 ?

decentral1se commented:
I'm not sure but you can quickly find out on the #ansible-dev channel or it is probably documented somewhere on the ROADMAP* files in the documentation. As the `core_review` label is on this now, I won't be able to get this merged from the Linode team side. So, your best bet is to push for core devs (ones commenting here and some on #ansible-dev channel) to get it in. Hope that helps! Thanks for the work.

mcg commented:
@decentral1se Thanks again for the efforts and help.

bcoca reviewed (changes requested):
use compose instead of having to add specific options per connection variable

mcg committed:
Removes custom instance_access variable

This allows the functionality of the custom instance_access variable by using
compose instead.

s-hertel commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -39,6 +41,8 @@
           default: []
           type: list
           required: false
+    strict:
```
These are already included via the `constructed` fragment. You can remove these two lines and check by using `ansible-doc -t inventory linode` on your branch.
```suggestion

```

s-hertel commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -39,6 +41,8 @@
           type: list
           required: false
+    strict:
+    compose:
```
```suggestion

```

s-hertel commented on lib/ansible/plugins/inventory/linode.py:
```diff
@@ -52,13 +56,16 @@
 '''
 
 import os
+import ipaddress
```
This should probably have a documented requirement or you should fall back to using `ansible.module_utils.compat.ipaddress` if there's an ImportError

mkrizek commented:
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
